File: Programmers/C++/순위 검색.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#define FOR(i,n) for(int i=0;i<(n);++i)
#define FORS(i,s,n) for(int i=(s);i<(n);++i)

using namespace std;
vector<int> res;
unordered_map<string,vector<int>> um;

vector<int> solution(vector<string> info, vector<string> query) {
    for(string s: info) {
        vector<string> v;
        string tmp = "";
        FOR(i,s.size()+1) {
            if(i==s.size() || s[i]==' ') {
                v.push_back(tmp);
                tmp="";
            } else tmp+=s[i];
        }
        um["----"].push_back(stoi(v[4]));
        
        um["-"+v[1]+v[2]+v[3]].push_back(stoi(v[4]));
        um[v[0]+"-"+v[2]+v[3]].push_back(stoi(v[4]));
        um[v[0]+v[1]+"-"+v[3]].push_back(stoi(v[4]));
        um[v[0]+v[1]+v[2]+"-"].push_back(stoi(v[4]));
        
        um["--"+v[2]+v[3]].push_back(stoi(v[4]));
        um["-"+v[1]+"-"+v[3]].push_back(stoi(v[4]));
        um["-"+v[1]+v[2]+"-"].push_back(stoi(v[4]));
        um[v[0]+"--"+v[3]].push_back(stoi(v[4]));
        um[v[0]+"-"+v[2]+"-"].push_back(stoi(v[4]));
        um[v[0]+v[1]+"--"].push_back(stoi(v[4]));
        
        um["---"+v[3]].push_back(stoi(v[4]));
        um["--"+v[2]+"-"].push_back(stoi(v[4]));
        um["-"+v[1]+"--"].push_back(stoi(v[4]));
        um[v[0]+"---"].push_back(stoi(v[4]));
        
        um[v[0]+v[1]+v[2]+v[3]].push_back(stoi(v[4]));
    }
    
    for(string s: query) {
        vector<string> v;
        string tmp = "";
        FOR(i,s.size()+1) {
            if(i==s.size() || s[i]==' ') {
                v.push_back(tmp);
                tmp="";
            } else tmp+=s[i];
        }
        string cur = "";
        FOR(i,v.size()-1) {
            string x = v[i];
            if(x=="and") continue;
            cur+=x;
        }
        int score = stoi(v[v.size()-1]);
        int cnt = 0;
        for(int k: um[cur]) if(k>=score) ++cnt;
        res.push_back(cnt);
    }
    return res;
}

File: Programmers/Swift/순위 검색.swift
import Foundation

var hash = [String : [Int]]()
var sorted = [String : Bool]()
var res = [Int]()

func solution(_ info:[String], _ query:[String]) -> [Int] {
    
    for str in info {
        var v = str.split(separator:" ").map{String($0)}
        let score = Int(v[4])!
        
        var str = String(v[0]+v[1]+v[2]+v[3])
        if hash[str] == nil {hash[str] = [Int]()}
        hash[str]!.append(score)
        
        str = String("-"+v[1]+v[2]+v[3])
        if hash[str] == nil {hash[str] = [Int]()}
        hash[str]!.append(score)
        str = String(v[0]+"-"+v[2]+v[3])
        if hash[str] == nil {hash[str] = [Int]()}
        hash[str]!.append(score)
        str = String(v[0]+v[1]+"-"+v[3])
        if hash[str] == nil {hash[str] = [Int]()}
        hash[str]!.append(score)
        str = String(v[0]+v[1]+v[2]+"-")
        if hash[str] == nil {hash[str] = [Int]()}
        hash[str]!.append(score)
        
        str = String("-"+"-"+v[2]+v[3])
        if hash[str] == nil {hash[str] = [Int]()}
        hash[str]!.append(score)
        str = String("-"+v[1]+"-"+v[3])
        if hash[str] == nil {hash[str] = [Int]()}
        hash[str]!.append(score)
        str = String("-"+v[1]+v[2]+"-")
        if hash[str] == nil {hash[str] = [Int]()}
        hash[str]!.append(score)
        str = String(v[0]+"-"+"-"+v[3])
        if hash[str] == nil {hash[str] = [Int]()}
        hash[str]!.append(score)
        str = String(v[0]+"-"+v[2]+"-")
        if hash[str] == nil {hash[str] = [Int]()}
        hash[str]!.append(score)
        str = String(v[0]+v[1]+"-"+"-")
        if hash[str] == nil {hash[str] = [Int]()}
        hash[str]!.append(score)
        
        str = String("-"+"-"+"-"+v[3])
        if hash[str] == nil {hash[str] = [Int]()}
        hash[str]!.append(score)
        str = String("-"+"-"+v[2]+"-")
        if hash[str] == nil {hash[str] = [Int]()}
        hash[str]!.append(score)
        str = String("-"+v[1]+"-"+"-")
        if hash[str] == nil {hash[str] = [Int]()}
        hash[str]!.append(score)
        str = String(v[0]+"-"+"-"+"-")
        if hash[str] == nil {hash[str] = [Int]()}
        hash[str]!.append(score)
        
        str = String("----")
        if hash[str] == nil {hash[str] = [Int]()}
        hash[str]!.append(score)
    }
    
    for str in query {
        var v = str.split(separator:" ").filter{$0 != "and"}
        
        let str = String(v[0]+v[1]+v[2]+v[3])
        if hash[str] == nil {
            res.append(0)
            continue
        }
        if sorted[str] == nil {
            hash[str]!.sort()
            sorted[str] = true
        }
        let num = Int(v[4])!
        
        //binary search
        var mx = -1
        var l = 0
        var r = hash[str]!.count-1
        while l <= r {
            let mid = (l+r)/2
            if hash[str]![mid] < num {
                mx = max(mx, mid)
                l = mid+1
            } else {
                r = mid-1
            }
        }
        res.append(hash[str]!.count-mx-1)
    }
    return res
}
